### PR TITLE
feat: define skeleton proto files

### LINKED ISSUE #508: Define x/oracle and x/datadeal with proto files
Based on a new `main` branch (branched from `release/v2.0.x`).

### PR DESCRIPTION
**Pull request by gyuguen**
Closes #508

### CODE REVIEW

File: proto/panacea/datadeal/v2/consent.proto
Comment (gyuguen) [inline proto/panacea/datadeal/v2/consent.proto:1]: I don't know this file name is correct.
certificate.go? consent.go? 
Please give me a good idea!

**Review by audtlr24 — APPROVED** [review]
lgtm!

File: proto/panacea/datadeal/v2/consent.proto
Comment (H4NLee) [inline proto/panacea/datadeal/v2/consent.proto:18]: why don't we use `oracle_address` to be clear?
I think we have many similar words can be confused: node, oracle, operator, validator, etc.
For oracle, it would be better to unify with the word `oracle`.

File: proto/panacea/datadeal/v2/deal.proto
Comment (H4NLee) [inline proto/panacea/datadeal/v2/deal.proto:19]: Do we still need this nonce? I think it was required to make consensus between oracles.

File: proto/panacea/datadeal/v2/tx.proto
Comment (H4NLee) [inline proto/panacea/datadeal/v2/tx.proto:29]: Same here :)
If needed, please ignore this comment as well.

File: proto/panacea/oracle/v2/oracle.proto
Comment (H4NLee) [inline proto/panacea/oracle/v2/oracle.proto:10]: How about `oracle_address` here too?

File: proto/panacea/oracle/v2/oracle.proto
Comment (H4NLee) [inline proto/panacea/oracle/v2/oracle.proto:11]: I think this status is not needed. 
If it is for upgrade, we can use `unique_id` to know whether the oracles have been upgraded

File: proto/panacea/oracle/v2/oracle.proto
Comment (H4NLee) [inline proto/panacea/oracle/v2/oracle.proto:48]: We may need to have status of `OracleRegistration`.
- pending : waiting for approval of oracle registration
- completed : received at least 1 approval of oracle registration

And also `encrypted_oracle_priv_key` may be required which is empty when `pending` status and updated when the first oracle approves.

**Review by H4NLee — COMMENTED** [review]
Left some comments :)

File: proto/panacea/datadeal/v2/deal.proto
Comment (audtlr24) [reply proto/panacea/datadeal/v2/deal.proto:19]: The nonce is required for `AES256` encryption. It is used to store the nonce used when performing encryption/decryption with combineKey. In deal, it is used to encrypt and decrypt the `certificate` with combineKey. It seems that there will be no problem if this value is `Params`, but in the previous implementation, each deal had a separate nonce.

File: proto/panacea/datadeal/v2/consent.proto
Comment (gyuguen) [reply proto/panacea/datadeal/v2/consent.proto:18]: Yes, We used a lot of words for Oracle's address. 
We set this name to `operator_address` in the architecture discussion.
But I think `oracle_address` is better. :)

File: proto/panacea/datadeal/v2/deal.proto
Comment (gyuguen) [reply proto/panacea/datadeal/v2/deal.proto:19]: And the CID of IPFS is always changed. Because encrypted data always changes when using a random nonce.

File: proto/panacea/oracle/v2/oracle.proto
Comment (gyuguen) [reply proto/panacea/oracle/v2/oracle.proto:48]: It is used only for Oracle register information. We don't manage register oracle status.
We can also get encrypted data from `MsgApproveOracleRegistration` tx.
New oracles need to subscirbe to `MsgApproveOracleRegistration` tx

File: proto/panacea/oracle/v2/oracle.proto
Comment (gyuguen) [reply proto/panacea/oracle/v2/oracle.proto:11]: You're right.
I think currently not required.
I will remove soon.

Comment (gyuguen) [reply proto/panacea/oracle/v2/oracle.proto]: fixed: https://github.com/medibloc/panacea-core/pull/515/commits/8eccb41513d2d18b87ff343c73155def0844c388

File: proto/panacea/datadeal/v2/consent.proto
Comment (gyuguen) [reply proto/panacea/datadeal/v2/consent.proto]: fixed: https://github.com/medibloc/panacea-core/pull/515/commits/8eccb41513d2d18b87ff343c73155def0844c388

File: proto/panacea/datadeal/v2/deal.proto
Comment (H4NLee) [reply proto/panacea/datadeal/v2/deal.proto:19]: > The nonce is required for AES256 encryption. It is used to store the nonce used when performing encryption/decryption with combineKey. In deal, it is used to encrypt and decrypt the certificate with combineKey. It seems that there will be no problem if this value is Params, but in the previous implementation, each deal had a separate nonce.

We can encrypt/decrypt data using `AES256` with randomly generated nonce.
Since the data is validated by the only one oracle, I think the nonce is not required.
In doracle, if we don't use nonce, all the oracles encrypt the data using each nonce, leading to each different value of `encrypted_oracle_priv_key` and consensus can't be made.
However in this scalable oracle, we don't need any consensus so that I thought the randomly generated nonce can be used. If I miss something, please let me know :pray:

Comment (H4NLee) [reply proto/panacea/datadeal/v2/deal.proto:19]: > And the CID of IPFS is always changed. Because encrypted data always changes when using a random nonce.

I don't understand the meaning of `the CID of IPFS is always changed`. Could you elaborate more??

Comment (gyuguen) [reply proto/panacea/datadeal/v2/deal.proto:19]: Yes, We can use a random nonce as you said.
However I'm worried that the data stored in IPFS keeps changing CID.(Even though it is the same data).

Comment (H4NLee) [reply proto/panacea/datadeal/v2/deal.proto:19]: Ah, I got your point.

Comment (H4NLee) [reply proto/panacea/datadeal/v2/deal.proto:19]: Right. As you said, we'd better to use nonce in encryption. Thank you all!

File: third_party/proto/cosmos/staking/v1beta1/staking.proto
Comment (H4NLee) [inline third_party/proto/cosmos/staking/v1beta1/staking.proto:89]: Oh, `staking.proto` has also been changed. Please check again :)

Comment (gyuguen) [reply third_party/proto/cosmos/staking/v1beta1/staking.proto]: Oh. I missed it. 
Thank you.

File: proto/panacea/oracle/v2/oracle.proto
Comment (H4NLee) [reply proto/panacea/oracle/v2/oracle.proto]: I got it. The `encrypted_oracle_priv_key` will not be stored, just subscribing the Tx.

Then, multiple `MsgApproveOracleRegistration`s from multiple oracles can be executed successfully because no status or state is stored, right?

Comment (gyuguen) [reply proto/panacea/oracle/v2/oracle.proto]: Exactly.

**Review by H4NLee — APPROVED** [review]
LGTM :+1:

**Review by inchori — APPROVED** [review]
LGTM 👍

**Review by youngjoon-lee — APPROVED** [review]

File: proto/panacea/datadeal/v2/consent.proto
Comment (gyuguen) [reply proto/panacea/datadeal/v2/consent.proto:1]: OK. I changed name to `certificate.proto`.
Thank you

File: proto/panacea/datadeal/v2/deal.proto
Comment (gyuguen) [reply proto/panacea/datadeal/v2/deal.proto]: Confirm. Remove the field 'nonce'.

**Pull request merged**
